### PR TITLE
add reporter to modules

### LINKED ISSUE #386: Add `teal.reporter` into tmg modules
- blocked by https://github.com/insightsengineering/teal.modules.clinical/issues/505

### PR DESCRIPTION
**Pull request by mhallal1**
closes #386 
dependency: https://github.com/insightsengineering/teal.reporter/pull/104

-Added reporter into all modules except `tm_front_page`, `tm_file_viewer` and `tm_data_table` as I did not see the necessity there, let me know if you disagree.
-`tm_scatterplot_matrix` plots were displayd as blank in the report, issue opened in `teal.reporter`: https://github.com/insightsengineering/teal.reporter/issues/98
-The bottom outliers table in the `tm_outliers` app is not rendered in the chunks therefore it is not included in the report (only the upper table and the plot).
-the reporter is also added in `tm_variable_browser` to add the plot, i am open to remove it if there is no need for it.

Please test with the example apps of each module

### CODE REVIEW

**Comment by Polkas** [comment]
Please take into account
https://github.com/insightsengineering/teal.reporter/pull/100
and
https://github.com/insightsengineering/teal.reporter/pull/97

**Comment by Polkas** [comment]
You need to update the teal.reporter staged.deps yaml too. This is needed as we report the up/down streams.

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # teal.modules.general 0.2.14.9003
 
+### Enhancements
+* Added the `teal.reporter` functionality to the modules.
```
Comment (Polkas) [inline NEWS.md:4]:
```suggestion
* Added the `teal.reporter` functionality to all modules.
```

File: R/tm_a_regression.R [hunk 2]
```diff
@@ -869,6 +879,36 @@ srv_a_regression <- function(id,
         format(code_header())
       )
     )
+
+    ### REPORTER
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Regression plot")
+        card$append_text("Regression plot", "header2")
```
Comment (Polkas) [inline R/tm_a_regression.R:888]:
```suggestion
        card$set_name("Linear Regression Plot")
        card$append_text("Linear Regression Plot", "header2")
```

File: R/tm_a_pca.R [hunk 3]
```diff
@@ -962,5 +971,39 @@ srv_a_pca <- function(id, datasets, dat, plot_height, plot_width, ggplot2_args)
       datanames = teal.transform::get_extract_datanames(list(dat)),
       modal_title = "R code for PCA"
     )
+
+    ### REPORTER
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("PCA plot")
+        card$append_text("PCA plot", "header2")
```
Comment (Polkas) [inline R/tm_a_pca.R:980]:
```suggestion
        card$set_name("PCA Plot")
        card$append_text("PCA Plot", "header2")
        card$append_text("Principal Component Analysis Plot", "header3")
```

File: R/tm_g_distribution.R [hunk 4]
```diff
@@ -1131,15 +1141,15 @@ srv_distribution <- function(id,
       rownames = FALSE
     )
 
-    teal.widgets::plot_with_settings_srv(
+    pws <- teal.widgets::plot_with_settings_srv(
```
Comment (Polkas) [inline R/tm_g_distribution.R:1144]:
```suggestion
    pws1 <- teal.widgets::plot_with_settings_srv(
```

File: R/tm_g_distribution.R [hunk 5]
```diff
@@ -1131,15 +1141,15 @@ srv_distribution <- function(id,
...
       id = "hist_plot",
       plot_r = dist_r,
       height = plot_height,
       width = plot_width,
       brushing = FALSE
     )
 
-    teal.widgets::plot_with_settings_srv(
+    pws1 <- teal.widgets::plot_with_settings_srv(
```
Comment (Polkas) [inline R/tm_g_distribution.R:1152]:
```suggestion
    pws2 <- teal.widgets::plot_with_settings_srv(
```

File: R/tm_g_distribution.R [hunk 6]
```diff
@@ -1160,6 +1170,48 @@ srv_distribution <- function(id,
...
+        card$set_name("Association plot")
+        card$append_text("Association plot", "header2")
+        card$append_text("Filter State", "header3")
+        card$append_fs(datasets$get_filter_state())
+        card$append_text("Main Element", "header3")
+        if (input$tabs == "Histogram") {
+          card$append_plot(dist_r(), dim = pws$dim())
+        } else {
+          card$append_plot(qq_r(), dim = pws1$dim())
+        }
```
Comment (Polkas) [inline R/tm_g_distribution.R:1187]:
```suggestion
        if (input$tabs == "Histogram") {
          card$append_plot(dist_r(), dim = pws1$dim())
        } else if (input$tabs == "QQplot") {
          card$append_plot(qq_r(), dim = pws2$dim())
        }
```

We persists the `else if` even in the main module code as a tab could be added ...

File: R/tm_g_scatterplot.R [hunk 7]
```diff
@@ -789,5 +799,36 @@ srv_g_scatterplot <- function(id,
...
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Scatter plot")
+        card$append_text("Scatter plot", "header2")
+        card$append_text("Filter State", "header3")
+        card$append_fs(datasets$get_filter_state())
+        card$append_text("Main Element", "header3")
+        card$append_plot(plot_r(), dim = brush$dim())
+
```
Comment (Polkas) [inline R/tm_g_scatterplot.R:813]:
```suggestion
```

File: R/tm_g_scatterplotmatrix.R [hunk 8]
```diff
@@ -322,6 +331,37 @@ srv_g_scatterplotmatrix <- function(id, datasets, variables, plot_height, plot_w
...
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Scatter plot matrix")
+        card$append_text("Scatter plot matrix", "header2")
+        card$append_text("Filter State", "header3")
+        card$append_fs(datasets$get_filter_state())
+        card$append_text("Main Element", "header3")
+        card$append_plot(plot_r(), dim = pws$dim())
+
```
Comment (Polkas) [inline R/tm_g_scatterplotmatrix.R:345]:
```suggestion
```

File: R/tm_g_scatterplotmatrix.R [hunk 9]
```diff
@@ -322,6 +331,37 @@ srv_g_scatterplotmatrix <- function(id, datasets, variables, plot_height, plot_w
       modal_title = show_r_code_title(),
       code_header = show_r_code_title()
     )
+
+    ### REPORTER
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Scatter plot matrix")
+        card$append_text("Scatter plot matrix", "header2")
```
Comment (Polkas) [inline R/tm_g_scatterplotmatrix.R:340]:
```suggestion
        card$set_name("Scatter Plot Matrix")
        card$append_text("Scatter Plot Matrix", "header2")
```

File: R/tm_missing_data.R [hunk 10]
```diff
@@ -1173,19 +1184,19 @@ srv_missing_data <- function(id, datasets, dataname, plot_height, plot_width, gg
       options = list(language = list(zeroRecords = "No variable selected"), pageLength = input$levels_table_rows)
     )
 
-    teal.widgets::plot_with_settings_srv(
+    pws1 <- teal.widgets::plot_with_settings_srv(
```
Comment (Polkas) [inline R/tm_missing_data.R:1187]: Here you nicely start form 1 with enumeration, we have to be consistent in other places.

File: R/tm_missing_data.R [hunk 11]
```diff
@@ -1197,5 +1208,46 @@ srv_missing_data <- function(id, datasets, dataname, plot_height, plot_width, gg
...
     )
+
+    ### REPORTER
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        summary_type <- input$summary_type
+        title <- paste0(summary_type, " plot")
+        card$set_name(paste0("Missing data - ", title))
+        card$append_text(title, "header2")
```
Comment (Polkas) [inline R/tm_missing_data.R:1219]: ?

File: R/tm_outliers.R [hunk 12]
```diff
@@ -1147,5 +1156,44 @@ srv_outliers <- function(id, datasets, outlier_var,
...
+        card$append_text("Main Element", "header3")
+        card$append_table(teal.code::chunks_get_var("summary_table", common_code_chunks()$common_stack))
+        if (tab_type == "Boxplot") {
+          card$append_plot(box_plot_plot_r(), dim = box_brush$dim())
+        } else if (tab_type == "Density plot") {
+          card$append_plot(density_plot_plot_r(), dim = density_brush$dim())
+        } else if (tab_type == "Cumulative distribution plot") {
+          card$append_plot(cumulative_plot_plot_r(), dim = cum_density_brush$dim())
+        }
+
```
Comment (Polkas) [inline R/tm_outliers.R:1178]:
```suggestion
```

File: R/tm_missing_data.R [hunk 13]
```diff
@@ -1197,5 +1208,46 @@ srv_missing_data <- function(id, datasets, dataname, plot_height, plot_width, gg
...
+        if (summary_type == "Summary") {
+          card$append_plot(summary_plot_r(), dim = pws1$dim())
+        } else if (summary_type == "Combinations") {
+          card$append_plot(combination_plot_r(), dim = pws2$dim())
+        } else if (summary_type == "By variable levels") {
+          card$append_table(teal.code::chunks_get_var("summary_data", table_chunks()))
+        } else if (summary_type == "Grouped by Subject") {
+          card$append_plot(by_subject_plot_r(), dim = pws3$dim())
+        }
+
```
Comment (Polkas) [inline R/tm_missing_data.R:1232]:
```suggestion
```

File: R/tm_variable_browser.R [hunk 14]
```diff
@@ -416,6 +425,29 @@ srv_variable_browser <- function(id, datasets, datasets_selected, ggplot2_args)
         }
       )
     })
+
+    ### REPORTER
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Variable browser plot")
+        card$append_text("Variable browser plot", "header2")
```
Comment (Polkas) [inline R/tm_variable_browser.R:434]:
```suggestion
        card$set_name("Variable Browser Plot")
        card$append_text("Variable Browser Plot", "header2")
```

File: R/tm_g_response.R [hunk 15]
```diff
@@ -436,5 +446,35 @@ srv_g_response <- function(id,
       datanames = teal.transform::get_extract_datanames(list(response, x, row_facet, col_facet)),
       modal_title = "R Code for Response Plot"
     )
+
+    ### REPORTER
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Response plot")
+        card$append_text("Response plot", "header2")
```
Comment (Polkas) [inline R/tm_g_response.R:455]:
```suggestion
        card$set_name("Response Plot")
        card$append_text("Response Plot", "header2")
```

File: R/tm_g_scatterplot.R [hunk 16]
```diff
@@ -789,5 +799,36 @@ srv_g_scatterplot <- function(id,
       modal_title = "R Code for a scatterplot",
       code_header = "Scatterplot"
     )
+
+    ### REPORTER
+    if (with_reporter) {
+      card_fun <- function(comment) {
+        card <- teal.reporter::TealReportCard$new()
+        card$set_name("Scatter plot")
+        card$append_text("Scatter plot", "header2")
```
Comment (Polkas) [inline R/tm_g_scatterplot.R:808]:
```suggestion
        card$set_name("Scatter Plot")
        card$append_text("Scatter Plot", "header2")
```

File: R/tm_missing_data.R [hunk 17]
```diff
@@ -230,6 +231,14 @@ encoding_missing_data <- function(id, summary_per_patient = FALSE, ggtheme) {
   ns <- NS(id)
 
   tagList(
+    ### Reporter
```
Comment (Polkas) [inline R/tm_missing_data.R:234]: It is not at the top

**Review by Polkas — COMMENTED** [review]
I do not feel comfortable with this direct calls to chunks, the effect is that we possibly could not take into account some use cases.

E.g. `teal.gallery::launch_app("exploratory")` where the outlier module is not working (card could not be added).

File: R/tm_missing_data.R (hunk 11 above)
Comment (Polkas) [reply R/tm_missing_data.R:1219]: to be consistent I will use it directly. I leave it to you.

Comment (mhallal1) [reply R/tm_missing_data.R:1219]: I use it twice so I prefer to define it and reuse it.

File: R/tm_missing_data.R (hunk 17 above)
Comment (mhallal1) [reply R/tm_missing_data.R:230]: should be fixed now :)

**Comment by mhallal1** [comment]
> I do not feel comfortable with this direct calls to chunks, the effect is that we possibly could not take into account some use cases.
> 
> E.g. `teal.gallery::launch_app("exploratory")` where the outlier module is not working (card could not be added).
> 
> Please think of some validation.

Yeah this stems from the fact that the `summary_tab` can be NULL when there are no outliers present. We check if it is `NULL` or not before adding it now.

File: R/tm_g_distribution.R [hunk 18]
```diff
@@ -1160,6 +1170,48 @@ srv_distribution <- function(id,
...
+        card$append_text("Filter State", "header3")
+        card$append_fs(datasets$get_filter_state())
+        card$append_text("Main Element", "header3")
+        if (input$tabs == "Histogram") {
+          card$append_plot(dist_r(), dim = pws1$dim())
+        } else if (input$tabs == "QQplot") {
+          card$append_plot(qq_r(), dim = pws2$dim())
+        }
+        card$append_text("Statistics table", "header3")
+        card$append_table(teal.code::chunks_get_var("summary_table", chunks = common_code_chunks()))
```
Comment (Polkas) [inline R/tm_g_distribution.R:1189]: When I run exploratory in teal.gallery, I got `Warning in chunks$get(var = var) :
  "summary_table" does not exist in the chunks.`.

Comment (mhallal1) [reply R/tm_g_distribution.R:1189]: `tryCatch` handles this now.

File: R/tm_a_regression.R [hunk 19]
```diff
@@ -888,7 +888,7 @@ srv_a_regression <- function(id,
         card$append_text("Linear Regression Plot", "header2")
         card$append_text("Filter State", "header3")
         card$append_fs(datasets$get_filter_state())
-        card$append_text("Main Element", "header3")
```
[the remark this answers is not in the document]
Comment (mhallal1) [reply R/tm_a_regression.R]: We use now `Plot` for plots and `Table` for tables.

**Review by Polkas — APPROVED** [review]
Fantastic work:) 🚀 🥇 

Any user related requirements could be easily added in future.

**Pull request merged**